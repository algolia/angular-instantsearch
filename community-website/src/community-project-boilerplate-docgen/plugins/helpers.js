const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
