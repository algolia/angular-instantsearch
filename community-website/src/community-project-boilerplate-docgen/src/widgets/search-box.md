---
title: Widgets - Search Box
layout: widget.pug
---

# Search Box

## Description

The Search Box component displays a search box that lets the user search for a specific query.

## Options

| Attribute          | Type      | Description
| -                  | -         | -
| `placeholder?`     | `string`  | The label of the input placeholder.
| `submitTitle?`     | `string`  | The submit button title.
| `resetTitle?`      | `string`  | The reset button title.
| `searchAsYouType?` | `boolean` | If you disable this option, new searches will only be triggered by clicking the search button or by pressing the enter key while the search box is focused.
| `autofocus?`       | `boolean` | Should the search box be focused on render?.

## Code example

```html
<ais-search-box placeholder='Search for products'>
</ais-search-box>
```
