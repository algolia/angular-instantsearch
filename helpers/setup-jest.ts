import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
