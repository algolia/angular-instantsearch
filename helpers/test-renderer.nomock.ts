import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
