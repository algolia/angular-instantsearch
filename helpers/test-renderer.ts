import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
