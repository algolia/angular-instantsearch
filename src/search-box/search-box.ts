import {
  Component,
  Input,
  Output,
  EventEmitter,
  Inject,
  PLATFORM_ID
} from "@angular/core";
import { connectSearchBox } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";

@Component({
  selector: "ng-ais-search-box",
  template: `
    <div [class]="cx()">
      <ng-ais-header [header]="header" [className]="cx('header')"></ng-ais-header>

      <div [class]="cx('body')">
        <form
          [class]="cx('form')"
          novalidate
          (submit)="handleSubmit($event)"
        >
          <input
            [class]="cx('input')"
            autocapitalize="off"
            autocorrect="off"
            placeholder="{{placeholder}}"
            role="textbox"
            spellcheck="false"
            type="text"
            [value]="state.query"
            (input)="handleChange($event.target.value)"
            (focus)="focus.emit($event)"
            (blur)="blur.emit($event)"
          />

          <button
            [class]="cx('submit')"
            type="submit"
            title="{{submitTitle}}"
            (click)="handleSubmit($event)"
          >
            <svg
              [ngClass]="cx('submitIcon')"
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 40 40"
              width="40"
              height="40"
            >
              <path d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z"></path>
            </svg>
          </button>

          <button
            [class]="cx('reset')"
            type="reset"
            title="{{resetTitle}}"
            (click)="handleReset($event)"
            style="display: none"
          >
            <svg
              [ngClass]="cx('resetIcon')"
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 20 20"
              width="20"
              height="20"
            >
              <path d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z"></path>
            </svg>
          </button>
        </form>
      </div>

      <ng-ais-footer [footer]="footer" [class]="cx('footer')"></ng-ais-footer>
    </div>
  `
})
export class NgAisSearchBox extends BaseWidget {
  @Input() public placeholder: string = "Search";
  @Input() public submitTitle: string = "Submit";
  @Input() public resetTitle: string = "Reset";
  @Input() public searchAsYouType: boolean = true;

  // Output events
  // form
  @Output() submit = new EventEmitter();
  @Output() reset = new EventEmitter();

  // input
  @Output() change = new EventEmitter();
  @Output() focus = new EventEmitter();
  @Output() blur = new EventEmitter();

  public state = {
    query: "",
    refine: noop
  };

  constructor(
    @Inject(PLATFORM_ID) public platformId: Object,
    searchInstance: NgAisInstance
  ) {
    super(searchInstance, "SearchBox");
    this.createWidget(connectSearchBox);
  }

  public handleChange(query: string) {
    this.change.emit(query);

    if (this.searchAsYouType) {
      this.state.refine(query);
    }
  }

  public handleSubmit(event: MouseEvent) {
    // send submit event to parent component
    this.submit.emit(event);

    event.preventDefault();

    if (!this.searchAsYouType) {
      this.state.refine(this.state.query);
    }
  }

  public handleReset(event: MouseEvent) {
    // send reset event to parent component
    this.reset.emit(event);

    // reset search
    this.state.refine("");
  }
}
