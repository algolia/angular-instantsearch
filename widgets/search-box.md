## `<ng-ais-search-box />`

The SearchBox component displays a search box that lets the user search for a specific query.

| Attribues         | Type      | Description
| -                 | -         | -
| `placeholder?`    | `string`  | The label of the input placeholder.
| `submitTitle`     | `string`  | The submit button title.
| `resetTitle`      | `string`  | The reset button title.
| `searchAsYouType` | `boolean` | If you disable this option, new searches will only be triggered by clicking the search button or by pressing the enter key while the search box is focused.