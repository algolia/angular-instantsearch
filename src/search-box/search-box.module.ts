import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisSearchBox } from "./search-box";

@NgModule({
  declarations: [NgAisSearchBox],
  entryComponents: [NgAisSearchBox],
  exports: [NgAisSearchBox],
  imports: [CommonModule]
})
export class NgAisSearchBoxModule {}
