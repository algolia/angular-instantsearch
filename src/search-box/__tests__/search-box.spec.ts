import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisSearchBox } from "../search-box";

const defaultState = {
  query: "foo",
  refine: jest.fn()
};

const render = createRenderer({
  defaultState,
  template: "<ais-search-box></ais-search-box>",
  TestedWidget: NgAisSearchBox
});

describe("SearchBox", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });
});
