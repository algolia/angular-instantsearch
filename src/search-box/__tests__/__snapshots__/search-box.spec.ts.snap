// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisSearchBox]}
>
  <ais-instantsearch>
    <ais-search-box>
      <div
        class="ais-SearchBox"
      >
        <form
          class="ais-SearchBox-form"
          novalidate=""
        >
          <input
            autocapitalize="off"
            autocorrect="off"
            class="ais-SearchBox-input"
            placeholder="Search"
            role="textbox"
            spellcheck="false"
            type="text"
          />
          <button
            class="ais-SearchBox-submit"
            title="Submit"
            type="submit"
          >
            <svg
              class="ais-SearchBox-submitIcon"
              height="40"
              ng-reflect-ng-class="ais-SearchBox-submitIcon"
              viewBox="0 0 40 40"
              width="40"
            >
              <path
                d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z"
              />
            </svg>
          </button>
          <button
            class="ais-SearchBox-reset"
            title="Reset"
            type="reset"
          >
            <svg
              class="ais-SearchBox-resetIcon"
              height="20"
              ng-reflect-ng-class="ais-SearchBox-resetIcon"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z"
              />
            </svg>
          </button>
        </form>
      </div>
    </ais-search-box>
  </ais-instantsearch>
</ng-component>
`;

exports[`SearchBox renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisSearchBox]}
>
  <ais-instantsearch>
    <ais-search-box>
      <div
        class="ais-SearchBox"
      >
        <form
          class="ais-SearchBox-form"
          novalidate=""
        >
          <input
            autocapitalize="off"
            autocorrect="off"
            class="ais-SearchBox-input"
            placeholder="Search"
            role="textbox"
            spellcheck="false"
            type="text"
          />
          <button
            class="ais-SearchBox-submit"
            title="Submit"
            type="submit"
          >
            <svg
              class="ais-SearchBox-submitIcon"
              height="40"
              ng-reflect-ng-class="ais-SearchBox-submitIcon"
              viewBox="0 0 40 40"
              width="40"
            >
              <path
                d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z"
              />
            </svg>
          </button>
          <button
            class="ais-SearchBox-reset"
            hidden=""
            title="Reset"
            type="reset"
          >
            <svg
              class="ais-SearchBox-resetIcon"
              height="20"
              ng-reflect-ng-class="ais-SearchBox-resetIcon"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z"
              />
            </svg>
          </button>
        </form>
      </div>
    </ais-search-box>
  </ais-instantsearch>
</ng-component>
`;
